db/db_impl/db_impl.cc

```diff
@@ -1828,126 +1764,350 @@ std::vector<Status> DBImpl::MultiGet(
   return stat_list;
 }
 
+template <class T, typename IterDerefFunc>
+bool DBImpl::MultiCFSnapshot(const ReadOptions& read_options,
```

Comment:
Consider to change to name from Snapshot, which can easily be confused as explicit RocksDB snapshot.
If my understanding is correct, T is used in two values, which indicates cf_list has one element and arbitrary ones. At the same time, they are guarded in an if-else. Can they be separated into two functions, so that we can get rid of this template, together with turning IterDerefFunc into a lamda?

MultiGet() with more than one CF can always go with the slow path. I don't think we need to optimize for that.

Reply:
Actually, T can be an ```autovector``` or a ```std::unordered_map```, which is why I made it a template argument so it'll be easier to iterate. We can use a lambda, but to avoid template the lambda cannot do any captures, which complicates things. The container of ```MultiGetColumnFamilyData```  will have to be passed as ```void *``` and typecast inside the lambda.

Reply:
I just noticed that Google C++ Style now has a section about type deduction: https://google.github.io/styleguide/cppguide.html#Type_deduction

Following the guide, the T should be OK (although maybe we shouldn't call it T but something else).

Regarding IterDerefFunc, I think lambda is better here because the type would be stronger. I agree that lambda is a little bit strange here, but template is not better. Can you explain why MultiGetColumnFamilyData will have to pass as `void*`? Just make the type `std::function<void(ColumnFamilyData*)>`. Will it work?

Or if you want to do the old style (there are still lots of code in RocksDB code, we can use function point.

Reply:
Yes, passing the lambda as ```std::function``` should work. Something like this -
```
<template MultiGetCFContainer>
DBImpl::MultiCFSnapshot(...,std::function<MultiGetColumnFamilyData*(MultiGetCFContainer::iterator&)>& iter_deref_func, ...)
```